File: ios/GhostChat/Core/Audio/SoundLibrary.swift
import AVFoundation
import Foundation

/// Plays short in-app sound effects. Sound files belong in `Resources/Sounds/` (CAF).
/// Missing files are logged-and-ignored rather than crashing — Phase 3 ships without assets.
final class SoundLibrary {

    enum Sound: String, CaseIterable {
        case ringtone
        case incomingMessage
        case sent
        case failed

        var fileName: String {
            switch self {
            case .ringtone:        return "ringtone.caf"
            case .incomingMessage: return "message-in.caf"
            case .sent:            return "message-out.caf"
            case .failed:          return "failed.caf"
            }
        }
    }

    private var players: [Sound: AVAudioPlayer] = [:]
    private var muted: Bool = false

    var isMuted: Bool { muted }

    init(muted: Bool = false) {
        self.muted = muted
    }

    func setMuted(_ value: Bool) {
        muted = value
        if value { stopAll() }
    }

    func play(_ sound: Sound, loop: Bool = false) {
        guard !muted else { return }
        do {
            let player = try makeOrGetPlayer(for: sound)
            player.numberOfLoops = loop ? -1 : 0
            player.currentTime = 0
            player.play()
        } catch {
            // Missing asset in Phase 3 — silent no-op.
        }
    }

    func stop(_ sound: Sound) {
        players[sound]?.stop()
    }

    func stopAll() {
        players.values.forEach { $0.stop() }
    }

    // MARK: - Private

    private func makeOrGetPlayer(for sound: Sound) throws -> AVAudioPlayer {
        if let cached = players[sound] { return cached }
        guard let url = Bundle.main.url(forResource: (sound.fileName as NSString).deletingPathExtension,
                                        withExtension: (sound.fileName as NSString).pathExtension) else {
            throw NSError(domain: "SoundLibrary", code: 404, userInfo: [NSLocalizedDescriptionKey: "missing \(sound.fileName)"])
        }
        let player = try AVAudioPlayer(contentsOf: url)
        player.prepareToPlay()
        players[sound] = player
        return player
    }
}

File: android/app/src/main/java/com/kordar/ghostchat/core/audio/SoundLibrary.kt
package com.kordar.ghostchat.core.audio

import android.content.Context
import android.media.MediaPlayer

/**
 * Plays short in-app sound effects. Assets live in `res/raw/` (ringtone.ogg etc.).
 * Missing files are logged-and-ignored — Phase 4 ships without assets committed
 * (placeholders in Phase 7).
 *
 * Mirror of iOS `SoundLibrary`.
 */
class SoundLibrary(
    private val context: Context,
    private val muted: () -> Boolean = { false }
) {
    enum class Sound(val resName: String) {
        Ringtone("ringtone"),
        IncomingMessage("message_in"),
        Sent("message_out"),
        Failed("failed")
    }

    private val players = mutableMapOf<Sound, MediaPlayer>()

    fun play(sound: Sound, loop: Boolean = false) {
        if (muted()) return
        val player = obtain(sound) ?: return
        player.isLooping = loop
        try {
            player.seekTo(0)
            player.start()
        } catch (_: IllegalStateException) {
            // Missing/broken file — silent no-op.
        }
    }

    fun stop(sound: Sound) {
        players[sound]?.takeIf { it.isPlaying }?.pause()
    }

    fun stopAll() {
        players.values.forEach { p -> if (p.isPlaying) p.pause() }
    }

    fun release() {
        players.values.forEach { it.release() }
        players.clear()
    }

    private fun obtain(sound: Sound): MediaPlayer? {
        players[sound]?.let { return it }
        val resId = context.resources.getIdentifier(sound.resName, "raw", context.packageName)
        if (resId == 0) return null
        return try {
            val p = MediaPlayer.create(context, resId) ?: return null
            players[sound] = p
            p
        } catch (_: Throwable) {
            null
        }
    }
}
